File: mapgen/navcheck.py
"""Graph-based connectivity validation for generated layouts.

BFS on the room-corridor topology, filtering corridors that aren't
physically navigable (too narrow, height difference too steep).
"""

from __future__ import annotations

from collections import deque
from dataclasses import dataclass, field

from .layout import (
    CORRIDOR_WIDTH,
    Layout,
    MAX_STAIR_RISE,
    STAIR_HEIGHT,
    WALKABLE_CLIMB,
    WALKABLE_RADIUS,
)

MIN_STEP_RUN = 16


@dataclass
class NavcheckResult:
    connected: bool = False
    component_count: int = 0
    unreachable_rooms: list[int] = field(default_factory=list)
    error: str = ""


def validate_layout_graph(layout: Layout) -> NavcheckResult:
    """Check that all rooms are reachable via navigable corridors."""
    result = NavcheckResult()
    n = len(layout.rooms)
    if n == 0:
        result.error = "No rooms"
        return result

    # Build adjacency list, filtering unnavigable corridors.
    adj: list[list[int]] = [[] for _ in range(n)]

    for c in layout.corridors:
        if c.room_a < 0 or c.room_a >= n or c.room_b < 0 or c.room_b >= n:
            continue

        # Corridor must be wide enough for player.
        width = (c.y1 - c.y0) if c.axis == "x" else (c.x1 - c.x0)
        if width < int(WALKABLE_RADIUS * 2):
            continue

        # Height difference must be climbable.
        dz = abs(c.z0_b - c.z0_a)
        if dz > 0:
            bridged = min(dz, MAX_STAIR_RISE)
            remainder = dz - bridged
            if remainder > WALKABLE_CLIMB:
                continue
            total_run = (c.x1 - c.x0) if c.axis == "x" else (c.y1 - c.y0)
            max_steps = total_run // MIN_STEP_RUN
            min_steps = (bridged + WALKABLE_CLIMB - 1) // WALKABLE_CLIMB
            if min_steps > max_steps:
                continue

        adj[c.room_a].append(c.room_b)
        adj[c.room_b].append(c.room_a)

    # BFS from room 0.
    visited = [False] * n
    q: deque[int] = deque([0])
    visited[0] = True

    while q:
        room = q.popleft()
        for neighbor in adj[room]:
            if not visited[neighbor]:
                visited[neighbor] = True
                q.append(neighbor)

    result.unreachable_rooms = [i for i in range(n) if not visited[i]]

    # Count connected components.
    result.component_count = 1
    for i in range(n):
        if visited[i]:
            continue
        result.component_count += 1
        cq: deque[int] = deque([i])
        visited[i] = True
        while cq:
            room = cq.popleft()
            for neighbor in adj[room]:
                if not visited[neighbor]:
                    visited[neighbor] = True
                    cq.append(neighbor)

    result.connected = len(result.unreachable_rooms) == 0
    return result

File: mapgen_cpp/src/navcheck.cpp
#include "mapgen/navcheck.h"

#include <cmath>
#include <queue>
#include <vector>

namespace mapgen {

NavcheckResult validate_layout_graph(const Layout& layout,
                                     const MapgenConfig& cfg) {
    NavcheckResult result;
    int n = static_cast<int>(layout.rooms.size());
    if (n == 0) {
        result.error = "No rooms";
        return result;
    }

    // Build adjacency list from corridors, filtering out physically
    // unnavigable connections.
    std::vector<std::vector<int>> adj(n);

    for (const auto& c : layout.corridors) {
        if (c.room_a < 0 || c.room_a >= n || c.room_b < 0 || c.room_b >= n)
            continue;

        // Check corridor is wide enough for player.
        int width = (c.axis == 'x') ? (c.y1 - c.y0) : (c.x1 - c.x0);
        if (width < static_cast<int>(cfg.navmesh.walkable_radius * 2))
            continue;

        // Check height difference is climbable (must match build_stairs logic).
        int dz = std::abs(c.z0_b - c.z0_a);
        if (dz > 0) {
            int bridged = std::min(dz, cfg.stairs.max_rise);
            int remainder = dz - bridged;
            int climb = static_cast<int>(cfg.navmesh.walkable_climb);
            // Remainder after stairs must be within a single step.
            if (remainder > climb)
                continue;
            int total_run = (c.axis == 'x') ? (c.x1 - c.x0) : (c.y1 - c.y0);
            int max_steps = total_run / cfg.stairs.min_step_run;
            int min_steps = (bridged + climb - 1) / climb;
            if (min_steps > max_steps)
                continue;
        }

        adj[c.room_a].push_back(c.room_b);
        adj[c.room_b].push_back(c.room_a);
    }

    // BFS from room 0.
    std::vector<bool> visited(n, false);
    std::queue<int> q;
    q.push(0);
    visited[0] = true;

    while (!q.empty()) {
        int room = q.front(); q.pop();
        for (int neighbor : adj[room]) {
            if (!visited[neighbor]) {
                visited[neighbor] = true;
                q.push(neighbor);
            }
        }
    }

    for (int i = 0; i < n; ++i)
        if (!visited[i])
            result.unreachable_rooms.push_back(i);

    // Count connected components.
    result.component_count = 1;
    for (int i = 0; i < n; ++i) {
        if (visited[i]) continue;
        ++result.component_count;
        std::queue<int> cq;
        cq.push(i);
        visited[i] = true;
        while (!cq.empty()) {
            int room = cq.front(); cq.pop();
            for (int neighbor : adj[room]) {
                if (!visited[neighbor]) {
                    visited[neighbor] = true;
                    cq.push(neighbor);
                }
            }
        }
    }

    result.connected = result.unreachable_rooms.empty();
    return result;
}

}  // namespace mapgen
